cluster/kubernetes/sync.go

```diff
@@ -2,36 +2,327 @@ package kubernetes
+			}
+
+			resourceClient := c.client.dynamicClient.Resource(schema.GroupVersionResource{
+				Group:    group,
+				Version:  version,
+				Resource: apiResource.Name,
+			})
+			data, err := resourceClient.List(listOptions)
+			if err != nil {
+				return nil, err
+			}
+
+			for i, item := range data.Items {
+				apiVersion := item.GetAPIVersion()
+				kind := item.GetKind()
+
+				itemDesc := fmt.Sprintf("%s:%s", apiVersion, kind)
+				// https://github.com/kontena/k8s-client/blob/6e9a7ba1f03c255bd6f06e8724a1c7286b22e60f/lib/k8s/stack.rb#L17-L22
+				if itemDesc == "v1:ComponentStatus" || itemDesc == "v1:Endpoints" {
+					continue
+				}
+				// TODO(michael) also exclude anything that has an ownerReference (that isn't "standard"?)
+
+				res := &kuberesource{obj: &data.Items[i]}
+				result[res.AssumedNamespaceResourceID(assumedNamespace).String()] = res
+			}
+		}
+	}
+
+	return result, nil
+}
+
+// exportResourcesInStack collates all the resources that belong to a
+// stack, i.e., were applied by flux.
+func (c *Cluster) getResourcesInStack(assumedNamespace string) (map[string]*kuberesource, error) {
+	return c.getResourcesBySelector(stackLabel, assumedNamespace) // means "has label <<stackLabel>>"
+}
+
+func applyMetadata(res resource.Resource, stack, checksum string) ([]byte, error) {
+	definition := map[interface{}]interface{}{}
+	if err := yaml.Unmarshal(res.Bytes(), &definition); err != nil {
+		return nil, errors.Wrap(err, fmt.Sprintf("failed to parse yaml from %s", res.Source()))
+	}
+
+	mixin := map[string]interface{}{}
+
+	if stack != "" {
+		mixinLabels := map[string]string{}
+		mixinLabels[stackLabel] = stack
+		mixin["labels"] = mixinLabels
+	}
+
+	if checksum != "" {
+		mixinAnnotations := map[string]string{}
+		mixinAnnotations[checksumAnnotation] = checksum
+		mixin["annotations"] = mixinAnnotations
+	}
+
+	mergo.Merge(&definition, map[interface{}]interface{}{
+		"metadata": mixin,
+	})
+
+	bytes, err := yaml.Marshal(definition)
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to serialize yaml after applying metadata")
+	}
+	return bytes, nil
+}
+
+// --- internal types for keeping track of syncing
+
+type applyObject struct {
+	ResourceID flux.ResourceID
+	Source     string
+	Payload    []byte
+}
+
 type changeSet struct {
-	objs map[string][]*apiObject
+	objs map[string][]applyObject
 }
 
 func makeChangeSet() changeSet {
-	return changeSet{objs: make(map[string][]*apiObject)}
+	return changeSet{objs: make(map[string][]applyObject)}
 }
 
-func (c *changeSet) stage(cmd string, o *apiObject) {
-	c.objs[cmd] = append(c.objs[cmd], o)
+func (c *changeSet) stage(cmd string, id flux.ResourceID, source string, bytes []byte) {
+	c.objs[cmd] = append(c.objs[cmd], applyObject{id, source, bytes})
 }
 
 // Applier is something that will apply a changeset to the cluster.
 type Applier interface {
 	apply(log.Logger, changeSet, map[flux.ResourceID]error) cluster.SyncError
+	getDefaultNamespace() string
```

Comment:
As I mentioned above, I think this belongs to the part of the code parsing resources (AFAICT, the fallback namespace is only passed to parsing functions)

Reply:
The reason it's here is that this is the interface wrapping `kubectl apply`, and the default namespace is essentially a property of `kubectl apply`.

The only reason we care about the default namespace is that when manifests that don't include one are applied, they will be defaulted. Adding it during parsing is a shortcut -- it's really part of the interpretation of the manifests as things to be applied, not inherent to the manifests themselves.